bundle/edu.gemini.itc.shared/src/main/scala/edu/gemini/itc/shared/ConfigExtractor.scala

```diff
@@ -0,0 +1,214 @@
+
+    def extractGuideStar(targets: GuideProbeTargets) =
+      targets.getPrimary.asScalaOpt.fold("No guide star selected".left[SPTarget])(_.right)
+
+    def extractMagnitude(guideStar: ITarget) = {
+      val r  = guideStar.getMagnitude(Band.r)
+      val R  = guideStar.getMagnitude(Band.R)
+      val UC = guideStar.getMagnitude(Band.UC)
+      if      (r.isDefined)  r.getValue.getBrightness.right
+      else if (R.isDefined)  R.getValue.getBrightness.right
+      else if (UC.isDefined) UC.getValue.getBrightness.right
+      else "No r, R or UC magnitude defined for guide star".left
+    }
+
+    if (altairIsPresent) {
+      for {
+        group     <- extractGroup
+        guideStar <- extractGuideStar(group)
+        magnitude <- extractMagnitude(guideStar.getTarget)
+        fieldLens <- extract[FieldLens]     (c, AoFieldLensKey)
+        wfsMode   <- extract[GuideStarType] (c, AoGuideStarTypeKey)
+      } yield {
+        val separation = distance(targetEnv.getBase, guideStar)
+        Some(AltairParameters(separation, magnitude, fieldLens, wfsMode))
+      }
+    } else {
+      None.right
+    }
+  }
+
+  private def extractObservingWavelength(c: Config): String \/ Double =
+    // Note: observing wavelength will only be available if instrument is configured for spectroscopy
+    if (c.containsItem(ObsWavelengthKey)) extractDoubleFromString(c, ObsWavelengthKey) else 0.0.right
+
+  // Extract a value of the given type from the configuration
+  private def extract[A](c: Config, key: ItemKey)(implicit clazz: ClassTag[A]): String \/ A =
+    extractWithThrowable[A](c, key).leftMap(_.getMessage)
+
+  // Extract a double value from a string in the configuration
+  private def extractDoubleFromString(c: Config, key: ItemKey): String \/ Double = {
+    val v = for {
+      s <- extractWithThrowable[String](c, key)
+      d <- \/.fromTryCatch(s.toDouble)
+    } yield d
+    v.leftMap(_.getMessage)
+   }
+
+  // Helper method that enforces that whatever we get from the config
+  // for the given key is not null and matches the type we expect.
+  private def extractWithThrowable[A](c: Config, key: ItemKey)(implicit clazz: ClassTag[A]): Throwable \/ A = {
+
+    def missingKey[A](key: ItemKey): \/[Throwable, A] =
+      new Error("Missing config value for key ${key.getPath}").left[A]
+
+    Option(c.getItemValue(key)).fold(missingKey[A](key)) { v =>
+      \/.fromTryCatch(clazz.runtimeClass.cast(v).asInstanceOf[A])
+    }
+
+  }
+
+  // Calculate distance between two coordinates in arc seconds
+  private def distance(t0: SPTarget, t1: SPTarget) = {
+    val c0 = toCoordinates(t0)
+    val c1 = toCoordinates(t1)
+    Coordinates.difference(c0, c1).distance.toArcsecs
+  }
+
+  // Turn SPTarget into new target model coordinates, note that DMS.getValue() guarantees value to be in [-90..90]
+  private def toCoordinates(t: SPTarget) = {
```

Comment:
A very similar method is present on `ModelConverters`, perhaps we could put it there